src/components/constraints/horizontal/horizontal.tsx

```diff
@@ -3,9 +3,9 @@ import classnames from 'classnames';
 import styles from './horizontal.mod.css';
 
 type HorizontalConstraint = 'xs' | 's' | 'm' | 'l' | 'xl' | 'scale';
-type HorizontalProps = {
-  constraint: HorizontalConstraint,
-};
+interface HorizontalProps {
```

Comment:
I think we should use `type` here. `interface` in TS is used when you want to be able to extend the interface. General rule of thumb is to only use interface if a consumer should be able to extend it. Given that we can `Exact`, `Diff` etc on `type` it is to my knowledge the preferred typing here.

Reply:
I thought so too, but tslint threw an error. I guess we can disable the rule then

Reply:
...have you, instead of defining our on TSLint config or copying it, into https://github.com/xojs/tslint-xo? I used XO a couple of times for JS and it lever let me down with bad defaults.

Reply:
How does it work? I never used it...

Reply:
Instead of having our own config in `tslint.json` we'd just 

```json
{
	"extends": "tslint-xo"
}
```

I also did not use it but have very good experience with xo for eslint. Maybe it's rather a follow up unless we have problem getting to a good config ourselves.

Reply:
I tried that and we get quite some warnings/errors. I would rather do this as a follow up once we have a setup running.

```bash
Warning: The 'no-unnecessary-type-assertion' rule requires type information.
Warning: The 'promise-function-async' rule requires type information.
Warning: The 'await-promise' rule requires type information.
Warning: The 'no-floating-promises' rule requires type information.
Warning: The 'no-for-in-array' rule requires type information.
Warning: The 'no-unnecessary-type-assertion' rule requires type information.
Warning: The 'no-use-before-declare' rule requires type information.
Warning: The 'restrict-plus-operands' rule requires type information.
Warning: The 'strict-type-predicates' rule requires type information.
Warning: The 'use-default-type-parameter' rule requires type information.
Warning: The 'deprecation' rule requires type information.
Warning: The 'promise-function-async' rule requires type information.
Warning: The 'await-promise' rule requires type information.
Warning: The 'no-unnecessary-type-assertion' rule requires type information.
Warning: The 'prefer-readonly' rule requires type information.
Warning: The 'no-unnecessary-type-assertion' rule requires type information.
Warning: The 'no-floating-promises' rule requires type information.
Warning: The 'no-for-in-array' rule requires type information.
Warning: The 'no-unnecessary-type-assertion' rule requires type information.
Warning: The 'promise-function-async' rule requires type information.
Warning: The 'await-promise' rule requires type information.
Warning: The 'promise-function-async' rule requires type information.
Warning: The 'await-promise' rule requires type information.
Warning: The 'no-floating-promises' rule requires type information.
Warning: The 'no-for-in-array' rule requires type information.
Warning: The 'no-floating-promises' rule requires type information.
Warning: The 'no-for-in-array' rule requires type information.
Warning: The 'promise-function-async' rule requires type information.
Warning: The 'no-unnecessary-qualifier' rule requires type information.
Warning: The 'await-promise' rule requires type information.
Warning: The 'no-use-before-declare' rule requires type information.
Warning: The 'restrict-plus-operands' rule requires type information.
Warning: The 'strict-type-predicates' rule requires type information.
Warning: The 'use-default-type-parameter' rule requires type information.
Warning: The 'deprecation' rule requires type information.
Warning: The 'no-use-before-declare' rule requires type information.
Warning: The 'prefer-readonly' rule requires type information.
Warning: The 'restrict-plus-operands' rule requires type information.
Warning: The 'strict-type-predicates' rule requires type information.
Warning: The 'no-floating-promises' rule requires type information.
Warning: The 'no-for-in-array' rule requires type information.
Warning: The 'no-use-before-declare' rule requires type information.
Warning: The 'restrict-plus-operands' rule requires type information.
Warning: The 'strict-type-predicates' rule requires type information.
Warning: The 'no-unnecessary-qualifier' rule requires type information.
Warning: The 'use-default-type-parameter' rule requires type information.
Warning: The 'deprecation' rule requires type information.
Warning: The 'use-default-type-parameter' rule requires type information.
Warning: The 'deprecation' rule requires type information.
Warning: The 'no-unnecessary-type-annotation' rule requires type information.
Warning: The 'prefer-readonly' rule requires type information.
 FAIL   tslint  types/jest.d.tslare' rule requires type information.
/Users/emmenko/dev/src/ui-kit/types/jest.d.ts:1:1
ERROR: 1:1  no-namespace  'namespace' and 'module' are disallowed
ERROR: 2:1  indent        tab indentation expected
ERROR: 3:1  indent        tab indentation expected
ERROR: 4:1  indent        tab indentation expected


 RUNS   tslint  types/index.d.ts
 RUNS   tslint  src/components/constraints/horizontal/horizontal.spec.tsx
 RUNS   tslint  src/components/constraints/horizontal/horizontal.tsx
 RUNS   tslint  src/test-utils.tsx
 RUNS   tslint  types/jest.d.ts

Test Suites: 503 passed, 503 of 508 total
Tests:       503 passed, 503 total
Snapshots:   0 total
Time:        3sWarning: The 'no-unnecessary-qualifier' rule requires type information.
The 'no-string-based-set-immediate' rule threw an error in '/Users/emmenko/dev/src/ui-kit/types/index.d.ts':
TypeError: Cannot read property 'kind' of undefined
    at GlobalReferenceCollector.SyntaxWalker.visitNode (/Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:276:22)
    at GlobalReferenceCollector.visitNode (/Users/emmenko/dev/src/ui-kit/node_modules/tslint-microsoft-contrib/utils/Scope.js:80:36)
    at Scope.addGlobalScope (/Users/emmenko/dev/src/ui-kit/node_modules/tslint-microsoft-contrib/utils/Scope.js:56:22)
    at NoStringParameterToFunctionCallWalker.ScopedSymbolTrackingWalker.visitModuleDeclaration (/Users/emmenko/dev/src/ui-kit/node_modules/tslint-microsoft-contrib/utils/ScopedSymbolTrackingWalker.js:100:20)
    at NoStringParameterToFunctionCallWalker.SyntaxWalker.visitNode (/Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:428:22)
    at /Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:535:63
    at visitNodes (/Users/emmenko/dev/src/ui-kit/node_modules/typescript/lib/typescript.js:16144:30)
    at Object.forEachChild (/Users/emmenko/dev/src/ui-kit/node_modules/typescript/lib/typescript.js:16372:24)
    at NoStringParameterToFunctionCallWalker.SyntaxWalker.walkChildren (/Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:535:12)
    at NoStringParameterToFunctionCallWalker.SyntaxWalker.visitSourceFile (/Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:228:14)
The 'no-string-based-set-interval' rule threw an error in '/Users/emmenko/dev/src/ui-kit/types/index.d.ts':
TypeError: Cannot read property 'kind' of undefined
    at GlobalReferenceCollector.SyntaxWalker.visitNode (/Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:276:22)
    at GlobalReferenceCollector.visitNode (/Users/emmenko/dev/src/ui-kit/node_modules/tslint-microsoft-contrib/utils/Scope.js:80:36)
    at Scope.addGlobalScope (/Users/emmenko/dev/src/ui-kit/node_modules/tslint-microsoft-contrib/utils/Scope.js:56:22)
    at NoStringParameterToFunctionCallWalker.ScopedSymbolTrackingWalker.visitModuleDeclaration (/Users/emmenko/dev/src/ui-kit/node_modules/tslint-microsoft-contrib/utils/ScopedSymbolTrackingWalker.js:100:20)
    at NoStringParameterToFunctionCallWalker.SyntaxWalker.visitNode (/Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:428:22)
    at /Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:535:63
    at visitNodes (/Users/emmenko/dev/src/ui-kit/node_modules/typescript/lib/typescript.js:16144:30)
    at Object.forEachChild (/Users/emmenko/dev/src/ui-kit/node_modules/typescript/lib/typescript.js:16372:24)
    at NoStringParameterToFunctionCallWalker.SyntaxWalker.walkChildren (/Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:535:12)
    at NoStringParameterToFunctionCallWalker.SyntaxWalker.visitSourceFile (/Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:228:14)
The 'no-string-based-set-timeout' rule threw an error in '/Users/emmenko/dev/src/ui-kit/types/index.d.ts':
TypeError: Cannot read property 'kind' of undefined
 PASS   tslint  types/index.d.tsSyntaxWalker.visitNode (/Users/emmenko/dev/src/ui-kit/node_modules/tslint/lib/language/walker/syntaxWalker.js:276:22)

 RUNS   tslint  types/index.d.ts
 RUNS   tslint  src/components/constraints/horizontal/horizontal.spec.tsx
 FAIL   tslint  src/components/constraints/horizontal/horizontal.spec.tsx
/Users/emmenko/dev/src/ui-kit/src/components/constraints/horizontal/horizontal.spec.tsx:2:8
ERROR: 2:8    object-curly-spacing  There should be no space after '{'
ERROR: 2:17   object-curly-spacing  There should be no space before '}'
ERROR: 8:1    indent                tab indentation expected
ERROR: 8:9    object-curly-spacing  There should be no space after '{'
ERROR: 8:23   object-curly-spacing  There should be no space before '}'
ERROR: 9:1    indent                tab indentation expected
ERROR: 9:28   quotemark             " should be '
ERROR: 10:1   indent                tab indentation expected
ERROR: 10:24  quotemark             " should be '
ERROR: 11:1   indent                tab indentation expected
ERROR: 12:1   indent                tab indentation expected
ERROR: 13:1   indent                tab indentation expected

 FAIL   tslint  src/components/constraints/horizontal/horizontal.tsx
/Users/emmenko/dev/src/ui-kit/src/components/constraints/horizontal/horizontal.tsx:6:6
ERROR: 6:6   interface-over-type-literal  Use an interface instead of a type literal.
ERROR: 7:1   indent                       tab indentation expected
ERROR: 8:2   semicolon                    Missing semicolon
ERROR: 11:1  indent                       tab indentation expected
ERROR: 12:1  indent                       tab indentation expected
ERROR: 13:1  indent                       tab indentation expected
ERROR: 14:1  indent                       tab indentation expected
ERROR: 15:1  indent                       tab indentation expected
ERROR: 16:1  indent                       tab indentation expected
ERROR: 17:1  indent                       tab indentation expected
ERROR: 18:1  indent                       tab indentation expected
ERROR: 19:1  indent                       tab indentation expected
ERROR: 20:1  indent                       tab indentation expected
ERROR: 21:1  indent                       tab indentation expected
ERROR: 22:1  indent                       tab indentation expected
ERROR: 23:1  indent                       tab indentation expected
ERROR: 24:1  indent                       tab indentation expected
ERROR: 25:1  indent                       tab indentation expected
ERROR: 26:1  indent                       tab indentation expected
ERROR: 29:7  variable-name                variable name must be in lowerCamelCase or UPPER_CASE
ERROR: 30:1  indent                       tab indentation expected
ERROR: 31:1  indent                       tab indentation expected
ERROR: 32:1  indent                       tab indentation expected
ERROR: 33:1  indent                       tab indentation expected
ERROR: 34:1  indent                       tab indentation expected
ERROR: 35:1  indent                       tab indentation expected
ERROR: 36:1  indent                       tab indentation expected
ERROR: 37:1  indent                       tab indentation expected
ERROR: 41:1  indent                       tab indentation expected

 FAIL   tslint  src/test-utils.tsx
/Users/emmenko/dev/src/ui-kit/src/test-utils.tsx:2:8
ERROR: 2:8    object-curly-spacing  There should be no space after '{'
ERROR: 2:17   object-curly-spacing  There should be no space before '}'
ERROR: 3:8    object-curly-spacing  There should be no space after '{'
ERROR: 3:38   object-curly-spacing  There should be no space before '}'
ERROR: 4:8    object-curly-spacing  There should be no space after '{'
ERROR: 4:17   object-curly-spacing  There should be no space before '}'
ERROR: 5:8    object-curly-spacing  There should be no space after '{'
ERROR: 5:30   object-curly-spacing  There should be no space before '}'
ERROR: 16:1   indent                tab indentation expected
ERROR: 17:1   indent                tab indentation expected
ERROR: 18:1   indent                tab indentation expected
ERROR: 18:22  comment-format        comment must start with uppercase letter
ERROR: 19:1   indent                tab indentation expected
ERROR: 20:1   indent                tab indentation expected
ERROR: 20:35  object-curly-spacing  There should be no space after '{'
ERROR: 20:61  object-curly-spacing  There should be no space before '}'
ERROR: 21:1   indent                tab indentation expected
ERROR: 22:1   indent                tab indentation expected
ERROR: 24:1   indent                tab indentation expected
ERROR: 25:1   indent                tab indentation expected
ERROR: 26:1   indent                tab indentation expected
ERROR: 27:1   indent                tab indentation expected
ERROR: 28:1   indent                tab indentation expected
ERROR: 29:1   indent                tab indentation expected
ERROR: 30:1   indent                tab indentation expected
ERROR: 30:5   comment-format        comment must start with uppercase letter
ERROR: 31:1   indent                tab indentation expected
ERROR: 31:5   comment-format        comment must start with uppercase letter
ERROR: 32:1   indent                tab indentation expected
ERROR: 32:5   comment-format        comment must start with uppercase letter
ERROR: 33:1   indent                tab indentation expected
ERROR: 35:3   comment-format        comment must start with uppercase letter
ERROR: 37:3   comment-format        comment must start with uppercase letter
ERROR: 38:8   object-curly-spacing  There should be no space after '{'
ERROR: 38:33  object-curly-spacing  There should be no space before '}'

Summary of all failing tests
 FAIL  types/jest.d.ts
/Users/emmenko/dev/src/ui-kit/types/jest.d.ts:1:1
ERROR: 1:1  no-namespace  'namespace' and 'module' are disallowed
ERROR: 2:1  indent        tab indentation expected
ERROR: 3:1  indent        tab indentation expected
ERROR: 4:1  indent        tab indentation expected

 FAIL  src/components/constraints/horizontal/horizontal.spec.tsx
/Users/emmenko/dev/src/ui-kit/src/components/constraints/horizontal/horizontal.spec.tsx:2:8
ERROR: 2:8    object-curly-spacing  There should be no space after '{'
ERROR: 2:17   object-curly-spacing  There should be no space before '}'
ERROR: 8:1    indent                tab indentation expected
ERROR: 8:9    object-curly-spacing  There should be no space after '{'
ERROR: 8:23   object-curly-spacing  There should be no space before '}'
ERROR: 9:1    indent                tab indentation expected
ERROR: 9:28   quotemark             " should be '
ERROR: 10:1   indent                tab indentation expected
ERROR: 10:24  quotemark             " should be '
ERROR: 11:1   indent                tab indentation expected
ERROR: 12:1   indent                tab indentation expected
ERROR: 13:1   indent                tab indentation expected

 FAIL  src/components/constraints/horizontal/horizontal.tsx
/Users/emmenko/dev/src/ui-kit/src/components/constraints/horizontal/horizontal.tsx:6:6
ERROR: 6:6   interface-over-type-literal  Use an interface instead of a type literal.
ERROR: 7:1   indent                       tab indentation expected
ERROR: 8:2   semicolon                    Missing semicolon
ERROR: 11:1  indent                       tab indentation expected
ERROR: 12:1  indent                       tab indentation expected
ERROR: 13:1  indent                       tab indentation expected
ERROR: 14:1  indent                       tab indentation expected
ERROR: 15:1  indent                       tab indentation expected
ERROR: 16:1  indent                       tab indentation expected
ERROR: 17:1  indent                       tab indentation expected
ERROR: 18:1  indent                       tab indentation expected
ERROR: 19:1  indent                       tab indentation expected
ERROR: 20:1  indent                       tab indentation expected
ERROR: 21:1  indent                       tab indentation expected
ERROR: 22:1  indent                       tab indentation expected
ERROR: 23:1  indent                       tab indentation expected
ERROR: 24:1  indent                       tab indentation expected
ERROR: 25:1  indent                       tab indentation expected
ERROR: 26:1  indent                       tab indentation expected
ERROR: 29:7  variable-name                variable name must be in lowerCamelCase or UPPER_CASE
ERROR: 30:1  indent                       tab indentation expected
ERROR: 31:1  indent                       tab indentation expected
ERROR: 32:1  indent                       tab indentation expected
ERROR: 33:1  indent                       tab indentation expected
ERROR: 34:1  indent                       tab indentation expected
ERROR: 35:1  indent                       tab indentation expected
ERROR: 36:1  indent                       tab indentation expected
ERROR: 37:1  indent                       tab indentation expected
ERROR: 41:1  indent                       tab indentation expected

 FAIL  src/test-utils.tsx
/Users/emmenko/dev/src/ui-kit/src/test-utils.tsx:2:8
ERROR: 2:8    object-curly-spacing  There should be no space after '{'
ERROR: 2:17   object-curly-spacing  There should be no space before '}'
ERROR: 3:8    object-curly-spacing  There should be no space after '{'
ERROR: 3:38   object-curly-spacing  There should be no space before '}'
ERROR: 4:8    object-curly-spacing  There should be no space after '{'
ERROR: 4:17   object-curly-spacing  There should be no space before '}'
ERROR: 5:8    object-curly-spacing  There should be no space after '{'
ERROR: 5:30   object-curly-spacing  There should be no space before '}'
ERROR: 16:1   indent                tab indentation expected
ERROR: 17:1   indent                tab indentation expected
ERROR: 18:1   indent                tab indentation expected
ERROR: 18:22  comment-format        comment must start with uppercase letter
ERROR: 19:1   indent                tab indentation expected
ERROR: 20:1   indent                tab indentation expected
ERROR: 20:35  object-curly-spacing  There should be no space after '{'
ERROR: 20:61  object-curly-spacing  There should be no space before '}'
ERROR: 21:1   indent                tab indentation expected
ERROR: 22:1   indent                tab indentation expected
ERROR: 24:1   indent                tab indentation expected
ERROR: 25:1   indent                tab indentation expected
ERROR: 26:1   indent                tab indentation expected
ERROR: 27:1   indent                tab indentation expected
ERROR: 28:1   indent                tab indentation expected
ERROR: 29:1   indent                tab indentation expected
ERROR: 30:1   indent                tab indentation expected
ERROR: 30:5   comment-format        comment must start with uppercase letter
ERROR: 31:1   indent                tab indentation expected
ERROR: 31:5   comment-format        comment must start with uppercase letter
ERROR: 32:1   indent                tab indentation expected
ERROR: 32:5   comment-format        comment must start with uppercase letter
ERROR: 33:1   indent                tab indentation expected
ERROR: 35:3   comment-format        comment must start with uppercase letter
ERROR: 37:3   comment-format        comment must start with uppercase letter
ERROR: 38:8   object-curly-spacing  There should be no space after '{'
ERROR: 38:33  object-curly-spacing  There should be no space before '}'


Test Suites: 4 failed, 504 passed, 508 total
Tests:       80 failed, 504 passed, 584 total
Snapshots:   0 total
Time:        3.901s
```

Reply:
Sure. It's also no problem to copy over whatever we think works well for us :)